src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -9228,5 +9158,565 @@ void M()
+  IL_007a:  ldstr      "" ""
+  IL_007f:  ldloc.s    V_6
+  IL_0081:  call       ""string string.Concat(string, string, string)""
+  IL_0086:  call       ""void System.Console.WriteLine(string)""
+  IL_008b:  ret
+}");
+        }
+
+        [Fact]
+        public void MixedDeclarationAndAssignmentCSharpEight()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        int x1;
+        (x1, string y1) = new C();
+        string y2;
+        (int x2, y2) = new C();
+    }
+
+    public void Deconstruct(out int a, out string b)
+    {
+        a = 1;
+        b = ""hello"";
+    }
+}
+";
+            CreateCompilation(source, parseOptions: TestOptions.Regular8).VerifyDiagnostics(
+                // (7,9): error CS8652: The feature 'Mixed declarations and expressions in deconstruction' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         (x1, string y1) = new C();
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "(x1, string y1) = new C()").WithArguments("Mixed declarations and expressions in deconstruction").WithLocation(7, 9),
+                // (9,9): error CS8652: The feature 'Mixed declarations and expressions in deconstruction' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         (int x2, y2) = new C();
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "(int x2, y2) = new C()").WithArguments("Mixed declarations and expressions in deconstruction").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void NestedMixedDeclarationAndAssignmentPermutations()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        int x1;
+        string y1;
+        (x1, (y1, var z1)) = new C();
+        System.Console.WriteLine(x1 + "" "" + y1 + "" "" + z1);
+
+        int x2;
+        bool z2;
+        (x2, (var y2, z2)) = new C();
+        System.Console.WriteLine(x2 + "" "" + y2 + "" "" + z2);
+
+        string y3;
+        bool z3;
+        (var x3, (y3, z3)) = new C();
+        System.Console.WriteLine(x3 + "" "" + y3 + "" "" + z3);
+
+        bool z4;
+        (var x4, (var y4, z4)) = new C();
+        System.Console.WriteLine(x4 + "" "" + y4 + "" "" + z4);
+
+        string y5;
+        (var x5, (y5, var z5)) = new C();
+        System.Console.WriteLine(x5 + "" "" + y5 + "" "" + z5);
+
+        int x6;
+        (x6, (var y6, var z6)) = new C();
+        System.Console.WriteLine(x6 + "" "" + y6 + "" "" + z6);
+
+        int x7;
+        (x7, var (y7, z7)) = new C();
+        System.Console.WriteLine(x7 + "" "" + y7 + "" "" + z7);
+    }
+
+    public void Deconstruct(out int a, out (string a, bool b) b)
+    {
+        a = 1;
+        b = (""hello"", true);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, parseOptions: TestOptions.RegularPreview, expectedOutput: @"1 hello True
```

Comment:
Could you add a C# 9 diagnostics check for nested case as well? Thanks

Reply:
Added here: https://github.com/dotnet/roslyn/pull/44476/files#diff-7c37d20e00ca6192dae6ed6ead852ab8R9857